src/librustc_mir/transform/const_prop.rs

```diff
@@ -553,19 +553,35 @@ impl<'mir, 'tcx> ConstPropagator<'mir, 'tcx> {
         right: &Operand<'tcx>,
         source_info: SourceInfo,
     ) -> Option<()> {
-        let r =
-            self.use_ecx(|this| this.ecx.read_immediate(this.ecx.eval_operand(right, None)?))?;
-        // Check for exceeding shifts *even if* we cannot evaluate the LHS.
-        if op == BinOp::Shr || op == BinOp::Shl {
-            // We need the type of the LHS. We cannot use `place_layout` as that is the type
-            // of the result, which for checked binops is not the same!
-            let left_ty = left.ty(&self.local_decls, self.tcx);
-            let left_size_bits = self.ecx.layout_of(left_ty).ok()?.size.bits();
-            let right_size = r.layout.size;
-            let r_bits = r.to_scalar().ok();
-            // This is basically `force_bits`.
-            let r_bits = r_bits.and_then(|r| r.to_bits_or_ptr(right_size, &self.tcx).ok());
-            if r_bits.map_or(false, |b| b >= left_size_bits as u128) {
+        if !right.needs_subst() {
```

Comment:
This seems entirely ad-hoc, and also not necessarily complete. What about `left`? What about the other functions that your previous PR exposed to subst-carrying data which were previously guarded by that `if` you moved around?

And what is even happening that we see uninitialized data, and that no assertion anywhere fires to detect that we run the engine on code it cannot run? This is the most concerning part to me. Cc @oli-obk

Reply:
> What about left?

Whoops, you're right. I've left left out.

> What about the other functions that your previous PR exposed to subst-carrying data which were previously guarded by that if you moved around?

The only reason I felt comfortable with doing this is `check_{unary, binary}_op` is only called from the `const_prop` method, so only `check_{unary, binary}_op` (and their callees) are newly exposed to subst-carrying data (speaking of which, I see `check_unary_op` also needs to be guarded because it evals its operand).

And what is even happening that we see uninitialized data, and that no assertion anywhere fires to detect that we run the engine on code it cannot run?

Per https://github.com/rust-lang/rust/issues/71353#issuecomment-616550763, isn't it just the error message raised during validation that's suspicious? I'm not sure that there's really uninitialised data here. I'll chuck a debugger on and see where the Err is being raised.

Reply:
> Per #71353 (comment), isn't it just the error message raised during validation that's suspicious? I'm not sure that there's really uninitialised data here. I'll chuck a debugger on and see where the Err is being raised.

That was my first theory. But in https://github.com/rust-lang/rust/issues/71353#issuecomment-616701709 it doesn't look like that any more.
Also, raw ptr consts which are NULL otherwise work fine. So, `ref_to_mplace` probably actually accepts NULL ptrs like it should.

> The only reason I felt comfortable with doing this is check_{unary, binary}_op is only called from the const_prop method, so only check_{unary, binary}_op (and their callees) are newly exposed to subst-carrying data (speaking of which, I see check_unary_op also needs to be guarded because it evals its operand).

If you guard `left` and `check_unary_op`, you fully un-did your previous PR and the bitshift warning will regress again. ;)

Reply:
Oh, I see where you are going... we can do the check for `left` *after* the bitshift check. Then the warning will still work fine and we fix the bug. Good idea.

However, we still should figure out what is happening that we see uninitialized data coming back here. That is *not* a good failure mode. Instead we should have gotten an ICE somewhere saying "there are still substs here, this const cannot be evaluated".

Reply:
> And what is even happening that we see uninitialized data, and that no assertion anywhere fires to detect that we run the engine on code it cannot run? This is the most concerning part to me. Cc @oli-obk

I'm entirely confused by this issue. I don't like the fix in this PR, it doesn't address the issue, it just makes sure we don't run into it.

It should be perfectly possible to evaluate `std::ptr::null()` for unknown (but `Sized`!) `T`. I don't see why it's failing validation of all things. I'd understand it if we'd get a `TooGeneric` error due to layout computation bailing out or sth. I think this is a bug in validation.

After checking what causes "uninitialized raw ptr", I believe the fault is at https://github.com/rust-lang/rust/blob/c6b55eed638db7db34e887ef255c765bc8910ec0/src/librustc_mir/interpret/place.rs#L307

well, not really, this is where I'd expect the `TooGeneric` error to come from, but `try_validation!` turns *every* error into a validation error: https://github.com/rust-lang/rust/blob/2dc5b602eee35d70e8e6e506a7ea07b6c7e0197d/src/librustc_mir/interpret/validity.rs#L54

We should probably not do that for any errors of the `InvalidProgramError` category: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/interpret/enum.InterpError.html#variant.InvalidProgram (and we can probably ICE for a few other categories)

Reply:
Ah, good point -- probably `try_validation!` turns `TooGeneric` into a validation failure.

I propose `try_validation!` should take a pattern for the error it expects, and then probably just propagate other errors through unchanged. (A fancy version of the macro would allow giving multiple patterns with different validation messages, like [what we do here](https://github.com/rust-lang/rust/blob/2dc5b602eee35d70e8e6e506a7ea07b6c7e0197d/src/librustc_mir/interpret/validity.rs#L362).)

Propagation is a bit risky because [of this catch here](https://github.com/rust-lang/rust/blob/2dc5b602eee35d70e8e6e506a7ea07b6c7e0197d/src/librustc_mir/interpret/validity.rs#L673) which catches errors not caught elsewhere in a very coarse-grained way... but the error variants are specific enough that this is probably okay.

If @oli-obk 's theory is right that will turn the weird validation failure into an ICE, which is an improvement.

Reply:
oh... I didn't know about that one, but I still think we should forward at least `InvalidProgram` to the validation caller instead of using `needs_subst`

Reply:
You [reviewed](https://github.com/rust-lang/rust/pull/69762) it, [though](https://github.com/rust-lang/rust/pull/69880). ;)

I agree it seems reasonable for validation to forward `InvalidProgram` instead of ICEing. Those are handled differently elsewhere anyway. At least in Miri; not sure about const-eval... I suppose it could `delay_span_bug` on any `InvalidProgram` error or so?

Reply:
const eval doesn't validate :) but `const_eval_validate` will use `struct_error` https://github.com/rust-lang/rust/blob/2dc5b602eee35d70e8e6e506a7ea07b6c7e0197d/src/librustc_mir/const_eval/eval_queries.rs#L216 which silently ignores `TooGeneric`

Reply:
We've very nearly come full circle https://github.com/rust-lang/rust/pull/70017#discussion_r392726654 :stuck_out_tongue_winking_eye:

Reply:
> `const_eval_validate`

Yeah that's what I meant.

>  silently ignores TooGeneric

Maybe that should be `delay_span_bug` instead.

Reply:
nono, that would be very wrong. We often invoke const eval on too generic code with `Reveal::UserFacing`, and we need to keep doing that. But for `Reveal::All` we can indeed report an error

Reply:
Good point, We should just forward it for now, then.

So, action items:
* Equip `try_validation!` with a pattern of errors to catch, the rest is forwarded. (Figuring out all the right patterns will take a few runs of the test suite here.)
* Forward `InvalidProgram` errors out of validation instead of ICEing.

Reply:
@jumbatm do you want to work on this alternative solution? (It's okay if you don't, in that case we'll look for someone else to do it. :) )

Reply:
I'm happy to keep going and work on the alternative solution. It definitely seems like the more robust way to fix the issue.

Reply:
Awesome, thanks. :)

Probably best to make that a new PR, as most comments here will not apply any more.